### PR TITLE
Create NFTLending interface, implement for NFTfi

### PR DESCRIPTION
**Pull request by wasabijk**

### CODE REVIEW

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/NFTLendingBase.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    /// @notice NFTLendingBase Constructor
+    /// @param _globalBNPL Global BNPL contract address
+    constructor(address _globalBNPL) {
+        require(_globalBNPL != address(0), "zero address");
+        globalBNPL = _globalBNPL;
+    }
+
+    /// @notice Update global BNPL contract address. NOTE: Only current global BNPL can call this function.
+    /// @param _globalBNPL New global BNPL contract address
+    function updateGlobalBNPL(address _globalBNPL) external onlyBNPL {
```
Comment (dev-wasabi) [inline contracts/lending/NFTLendingBase.sol:29]: If we have to update the global bnpl address, we will need to update for each contract. instead should we have a separate contract called `globalBNPLProvider` ?

File: contracts/lending/NFTFiLending.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        IERC721 nft = IERC721(offer.nftCollateralContract);
+        uint256 nftId = offer.nftCollateralId;
+
+        // Transfer NFT from BNPL contract
+        nft.safeTransferFrom(msg.sender, address(this), nftId);
+
+        // Approve
+        nft.setApprovalForAll(address(directLoanFixedCollectionOffer), true);
+
+        // Accetp offer on NFTfi
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:63]: nit: Accept

File: contracts/lending/NFTFiLending.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                    IDirectLoanFixedCollectionOffer.BorrowerSettings
+                )
+            );
+
+        IERC721 nft = IERC721(offer.nftCollateralContract);
+        uint256 nftId = offer.nftCollateralId;
+
+        // Transfer NFT from BNPL contract
+        nft.safeTransferFrom(msg.sender, address(this), nftId);
+
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:59]: Im thinking if this is gonna cost us even more gas. Like right now basically we will buy the NFT from the BNPL contract, transfer here and start a loan.

File: contracts/lending/NFTFiLending.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            loanTerms.maximumRepaymentAmount
+        );
+
+        // Pay back loan
+        directLoanFixedCollectionOffer.payBackLoan(loanId);
+
+        // Transfer collateral NFT to BNPL
+        IERC721(loanTerms.nftCollateralContract).safeTransferFrom(
+            address(this),
+            msg.sender,
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:105]: Let's pass this in as an argument. The repay function should take in the address that's gonna receive the collateral NFT

File: contracts/lending/NFTFiLending.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        // Transfer payment from BNPL
+        IERC20 token = IERC20(loanTerms.loanERC20Denomination);
+        token.safeTransferFrom(
+            msg.sender,
+            address(this),
+            loanTerms.maximumRepaymentAmount
+        );
+
+        // Approve token to `directLoanFixedCollectionOffer`
+        token.safeApprove(address(directLoanFixedCollectionOffer), 0);
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:93]: why is this for?

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/NFTFiLending.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:1]: The GlobalBNPL contract should use delegatecall for the borrow/repay functions. let's change the implementations to act accordingly

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/interfaces/IAddressProvider.sol [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    event NFTfiLendingUpdated(address indexed newAddress);
+    event X2Y2LendingUpdated(address indexed newAddress);
+
+    function getNFTfiLending() external view returns (address);
+
+    function setNFTfiLending(address _nftfiLending) external;
+
+    function getX2Y2Lending() external view returns (address);
+
+    function setX2Y2Lending(address _x2y2Lending) external;
```
Comment (dev-wasabi) [inline contracts/lending/interfaces/IAddressProvider.sol:14]: I don't think we need custom function for each lending, we would pass in a contract address or enum in the call which will determine which protocol to use

File: contracts/lending/NFTFiLending.sol (hunk 5 above)
Comment (dev-wasabi) [reply contracts/lending/NFTFiLending.sol:73]: ^

File: contracts/lending/NFTFiLending.sol [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        // Approve token to `directLoanFixedCollectionOffer`
+        IERC20 token = IERC20(loanTerms.loanERC20Denomination);
+        token.safeApprove(address(directLoanFixedCollectionOffer), 0);
+        token.safeApprove(
+            address(directLoanFixedCollectionOffer),
+            loanTerms.maximumRepaymentAmount
+        );
+
+        // Pay back loan
+        directLoanFixedCollectionOffer.payBackLoan(loanId);
```
Comment (dev-wasabi) [inline contracts/lending/NFTFiLending.sol:80]: are there instances in which we pay with just ETH? or is it always erc20?

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/NFTFiLending.sol (hunk 8 above)
Comment (wasabijk) [reply contracts/lending/NFTFiLending.sol:80]: it will be always ERC20

File: contracts/lending/NFTFiLending.sol (hunk 5 above)
Comment (wasabijk) [reply contracts/lending/NFTFiLending.sol:73]: when using `safeApprove`, we need to reset allowance to 0 first
this is why we are calling `safeApprove(0)`

File: contracts/lending/WasabiBNPL.sol [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    /// @notice Executes BNPL flow
+    /// @dev BNLP flow
+    ///      1. take flashloan
+    ///      2. buy nft from marketplace
+    ///      3. get loan from nft lending protocol
+    /// @param _value Call value
+    /// @param _marketplaceCallData List of marketplace calldata
+    /// @param _signatures Signatures
+    function bnpl(
+        bytes calldata _borrowData,
```
[the remark this answers is not in the document]
Comment (dev-wasabi) [reply contracts/lending/WasabiBNPL.sol:45]: for determining which nft to use for loan

File: contracts/lending/WasabiBNPL.sol [hunk 10]
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
```
Comment (dev-wasabi) [inline contracts/lending/WasabiBNPL.sol:1]: add unit tests

Comment (dev-wasabi) [reply contracts/lending/WasabiBNPL.sol:1]: add new settable variable `address factory` that will be used when calling WasabiOption#mint

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/WasabiBNPL.sol [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        // Buy NFT
+        bool marketSuccess = executeFunctions(_marketplaceCallData);
+        if (!marketSuccess) {
+            revert FunctionCallFailed();
+        }
+
+        (bool success, bytes memory result) = _nftLending.delegatecall(
+            abi.encodeWithSignature("borrow(bytes)", _borrowData)
```
Comment (dev-wasabi) [inline contracts/lending/WasabiBNPL.sol:69]: INFTLending.borrow.selector

Comment (wasabijk) [reply contracts/lending/WasabiBNPL.sol:69]: updated!

File: contracts/lending/interfaces/INFTLending.sol [hunk 12]
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+/// @notice NFTLending Interface
+interface INFTLending {
```
Comment (dev-wasabi) [inline contracts/lending/interfaces/INFTLending.sol:5]: let's add a new view function that returns the nft address and the nft id given a loan id

function getNFTDetails(uint256 _loanId) returns (address nftAddress, uint256 tokenId)

File: contracts/lending/WasabiBNPL.sol [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+        if (!addressProvider.isLending(_nftLending)) {
+            revert InvalidParam();
+        }
+
+        uint256 balanceBefore = address(this).balance;
+        if (balanceBefore < _value) {
+            revert InsufficientBalance();
+        }
+        balanceBefore -= msg.value;
```
Comment (dev-wasabi) [inline contracts/lending/WasabiBNPL.sol:80]: I think this should go like this:

```
uint256 balanceBefore = address(this).balance -= msg.value;
if (balanceBefore < _flashLoanAmount) {
    revert InsufficientBalance();
}
```

File: contracts/lending/interfaces/ILendingAddressProvider.sol [hunk 14]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
```
Comment (dev-wasabi) [inline contracts/lending/interfaces/ILendingAddressProvider.sol:1]: let's also have a function to remvoe addresses

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/Flashloan.sol [hunk 15]
```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./interfaces/IFlashloan.sol";
+
+contract Flashloan is IFlashloan, Ownable {
```
Comment (dev-wasabi) [inline contracts/lending/Flashloan.sol:8]: Let's add the following functions:

withdrawETH(uint256 _amount) onlyOwner

withdrawERC20(address contractAddress, uint256_ amount) onlyOwner

withdrawERC721(address contractAddress, uint256_ amount) onlyOwner

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/interfaces/x2y2/IXY3.sol [hunk 16]
```diff
@@ -2,51 +2,72 @@
...
     function borrow(
-        Offer calldata _offer,
+        Offer memory _offer,
         uint256 _nftId,
-        bool _isCollectionOffer,
-        Signature calldata _lenderSignature,
-        Signature calldata _brokerSignature,
-        CallData calldata _extraDeal
+        BrokerSignature memory _brokerSignature,
+        CallData memory _extraData
```
Comment (dev-wasabi) [inline contracts/lending/interfaces/x2y2/IXY3.sol]: shouldnt these be calldata?

File: contracts/lending/interfaces/x2y2/IXY3.sol [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+    function loanDetails(
+        uint32 _loanId
+    ) external view returns (LoanDetail memory);
+
+    function borrow(
+        Offer memory _offer,
+        uint256 _nftId,
+        BrokerSignature memory _brokerSignature,
+        CallData memory _extraData
```
Comment (dev-wasabi) [inline contracts/lending/interfaces/x2y2/IXY3.sol:70]: shouldnt these be calldata?

File: contracts/lending/X2Y2Lending.sol [hunk 18]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
```
Comment (dev-wasabi) [inline contracts/lending/X2Y2Lending.sol:1]: this one is not changed

File: contracts/lending/Flashloan.sol (hunk 15 above)
Comment (dev-wasabi) [reply contracts/lending/Flashloan.sol:8]: this too please

**Review by dev-wasabi — CHANGES_REQUESTED** [review]

File: contracts/lending/interfaces/x2y2/IXY3.sol (hunk 17 above)
Comment (dev-wasabi) [reply contracts/lending/interfaces/x2y2/IXY3.sol:70]: actully nvm

**Review by dev-wasabi — APPROVED** [review]

**Pull request merged**